pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -218,21 +218,30 @@ async def _prepare_request_parameters(
     ctx: GraphRunContext[GraphAgentState, GraphAgentDeps[DepsT, NodeRunEndT]],
 ) -> models.ModelRequestParameters:
     """Build tools and create an agent model."""
-    function_tool_defs: list[ToolDefinition] = []
+    function_tool_defs: dict[str, ToolDefinition] = {}
 
     run_context = build_run_context(ctx)
 
     async def add_tool(tool: Tool[DepsT]) -> None:
         ctx = run_context.replace_with(retry=tool.current_retry, tool_name=tool.name)
         if tool_def := await tool.prepare_tool_def(ctx):
-            function_tool_defs.append(tool_def)
+            # prepare_tool_def may change tool_def.name
+            if tool_def.name in function_tool_defs:
+                raise exceptions.UserError(
+                    f"Tool '{tool}' defines a tool whose name conflicts with existing tool: {tool_def.name!r}."
+                )
+            function_tool_defs[tool_def.name] = tool_def
 
     async def add_mcp_server_tools(server: MCPServer) -> None:
         if not server.is_running:
             raise exceptions.UserError(f'MCP server is not running: {server}')
         tool_defs = await server.list_tools()
-        # TODO(Marcelo): We should check if the tool names are unique. If not, we should raise an error.
-        function_tool_defs.extend(tool_defs)
+        for tool_def in tool_defs:
+            if tool_def.name in function_tool_defs:
+                raise exceptions.UserError(
+                    f"MCP Server '{server}' defines a tool whose name conflicts with existing tool: {tool_def.name!r}. Consider using `tool_prefix` to avoid name conflicts."
```

Comment:
@Wh1isper What exactly does `{server}` show up as here? Can we show the name or some other (presumed to be) unique identifier?

Reply:
Like `'MCPServerStdio(command='python', args=['-m', 'tests.mcp_server'], env=None, log_level=None, cwd=None, tool_prefix=None)'`

As `MCPServer` have no name, it's hard to represent. We can implement `__repr__` for http url or command and args.

Reply:
Does `MCP Server 'MCPServerStdio(command=python -m tests.mcp_server)' defines a tool whose name conflicts with existing tool: 'get_none'. Consider using `tool_prefix` to avoid name conflicts.` looks good? 

In general we don't want to expose environment variables or header.

Reply:
Currently expose `command`, `args` and `tool_prefix`  in error message

```
MCPServerStdio(command='python', args=['-m', 'tests.mcp_server'], tool_prefix=None)
```